### ISSUE TITLE
"Sort by likes" button in Agriculture and Air Quality topics does not sort correctly

### ISSUE THREAD

**Issue by khadija-nur** [publiclab/plots2#8598]
### _Sort by likes_ does not work in agriculture and air quality topics

- When you go to "https://publiclab.org/tag/air-quality" or "https://publiclab.org/tag/agriculture" 
and click on _sort by_ drop down menu then filter _by likes_ button the posts appear randomly 
instead of being in a descending order depending on the number of likes they have

- see below: 
- After filtering, the first post has 11 likes, the third one has 7 and the fourth has 8 likes

![like-11](https://user-images.githubusercontent.com/59174800/96297972-9a922e00-0fe9-11eb-8f3c-99a473f65159.PNG)

![likes-7](https://user-images.githubusercontent.com/59174800/96298004-a54cc300-0fe9-11eb-840b-a7b926412c71.PNG)

![likes-8](https://user-images.githubusercontent.com/59174800/96298018-aa117700-0fe9-11eb-9450-7a58debe0f7e.PNG)

### Please show us where to look

https://publiclab.org/tag/agriculture?order=likes
https://publiclab.org/tag/air-quality?order=likes

### What's your PublicLab.org username?
username: knur

### Browser, version, and operating system
- Windows 10 OS
- Chrome browser

****

**Comment by alvesitalo** [comment]
Hey, can I work on this?

**Comment by cesswairimu** [comment]
Hi @alvesitalo, we would love your help. Go ahead . Thanks

**Comment by ebarry** [comment]
Thank you for reporting this @khadija-nur -- this is a valuable contribution! 🤝
and thank you @@alvesitalo for putting in some work to fix this 😃

**Comment by alvesitalo** [comment]
Hi everyone, I tested locally with some posts but I can't see any problem with sorting by likes. Have you experienced this issue? @ebarry @cesswairimu

**Comment by cesswairimu** [comment]
@alvesitalo yeah, it is happening in production if you go to https://publiclab.org/tag/agriculture" and select sort by likes you will see the second and third post may have more likes than the first one. I will test locally see if I can replicate

**Comment by jywarren** [comment]
Hi all, thanks @khadija-nur for finding this, and @alvesitalo for looking deeper - it seems like it may be pretty intermittent, which can be really tough to debug. 

I wonder if we are sorting by cached likes count versus a count "on the fly" - see the model for tracking and caching likes at:

https://github.com/publiclab/plots2/blob/27a3839154e0cec071860448f99697f9c831042c/app/models/node.rb#L1039-L1046

The page itself is using `cached_likes` here:

https://github.com/publiclab/plots2/blob/27a3839154e0cec071860448f99697f9c831042c/app/controllers/tag_controller.rb#L121-L122

But see how the template is using `node.likers.count` ? Which may cause another database query and not be very efficient.

https://github.com/publiclab/plots2/blob/27a3839154e0cec071860448f99697f9c831042c/app/views/notes/_card.html.erb#L55

`node.likers` is defined here, and seems to use a different query... than the `cached_likes` code. Could they be drifting apart because some folks may like, but then get banned? In any case we should be able to use `node.cached_likes` instead in the template. That should be a pretty simple fix for now!

https://github.com/publiclab/plots2/blob/27a3839154e0cec071860448f99697f9c831042c/app/models/node.rb#L256-L264

**Comment by jywarren** [comment]
I'll spin out that last idea about filtering banned users from `cached_likes` into its own issue. It's just a one-line change too!

**Comment by jywarren** [comment]
I spun that out into a first-timers-only issue at https://github.com/publiclab/plots2/issues/8639 !

I could do the same for the template change above too, if it's helpful?

**Comment by alvesitalo** [comment]
Cool, it is. Thank you for clarifying this bug @jywarren

**Comment by jywarren** [comment]
I just did that over at #8691 - thanks @alvesitalo!

**Comment by gauravahlawat81** [comment]
Hey @jywarren . I was trying to reproduce this issue in my development environment, but I ran into some unexpected error. I can't find the nav link to 'Topics' when I try to run it locally.
![Screenshot from 2020-12-08 21-32-08](https://user-images.githubusercontent.com/36696399/101507959-23e83000-399d-11eb-928a-70c860506130.png)

**Comment by gauravahlawat81** [comment]
Hey @jywarren , I was trying to reproduce this error in my local environment, but I couldn't do it. The post were all sorted correctly and I observed the same [here](https://publiclab.org/tag/agriculture?order=likes), I wonder if this issue is solved now or are am I missing something else here ? :thinking:
